### PR TITLE
Added support for a pinned floating SliverAppBar

### LINKED ISSUE #8157: SliverAppBar does not support title bar and bottom bar with different behaviors
@Hixie 

When using a SliverAppBar in CustomScrollView (in the body attribute of the Scaffold), is it possible to have the title bar and the bottom bar with different behaviors? Say I would to have the title bar floating and the bottom bar pinned.
 
See [ #5126](https://github.com/flutter/flutter/issues/5126).

Thanks :)

### PR DESCRIPTION
**Pull request by HansMuller**
If SliverAppbar's pinned and floating parameters are true, then when scrolling upwards the appbar shrinks until only the it's bottom widget is visible. After that the bottom widget remains pinned.

Fixes https://github.com/flutter/flutter/issues/8157.

### CODE REVIEW

File: packages/flutter/lib/src/material/app_bar.dart [hunk 1]
```diff
@@ -93,6 +93,32 @@ class _ToolbarLayout extends MultiChildLayoutDelegate {
...
+// Bottom justify the kToolbarHeight child which may overflow the top.
+class _ToolbarContainerLayout extends SingleChildLayoutDelegate {
+  @override
+  BoxConstraints getConstraintsForChild(BoxConstraints constraints) {
+    return new BoxConstraints(
+      minWidth: constraints.minWidth,
+      maxWidth: constraints.maxWidth,
+      minHeight: kToolbarHeight,
+      maxHeight: kToolbarHeight,
+    );
```
Comment (abarth) [inline packages/flutter/lib/src/material/app_bar.dart]: `return constraings.tighten(height: kToolbarHeight)`

File: packages/flutter/lib/src/material/app_bar.dart [hunk 2]
```diff
@@ -93,6 +93,32 @@ class _ToolbarLayout extends MultiChildLayoutDelegate {
...
+  }
+
+  @override
+  Size getSize(BoxConstraints constraints) {
+    return constraints.tighten(height: kToolbarHeight).biggest;
```
Comment (abarth) [inline packages/flutter/lib/src/material/app_bar.dart]: Easier: `return new Size(constraings.maxWidth, kToolbarHeight)`

File: packages/flutter/lib/src/material/app_bar.dart [hunk 3]
```diff
@@ -422,24 +448,32 @@ class _AppBarState extends State<AppBar> {
...
-        context: context,
-        data: appBarIconTheme,
-        child: new DefaultTextStyle(
-          style: sideStyle,
-          child: toolbar,
+    // If the toolbar is allocated less than kToolbarHeight make it
+    // appear to scroll upwards within its shrinking container.
+    Widget appBar = new ClipRect(
+      child: new CustomSingleChildLayout(
+        delegate: new _ToolbarContainerLayout(),
```
Comment (abarth) [inline packages/flutter/lib/src/material/app_bar.dart]: Can we make _ToolbarContainerLayout const constructable and use `const _ToolbarContainerLayout()` here?

Comment (abarth) [inline packages/flutter/lib/src/material/app_bar.dart]: Looks like not.

File: packages/flutter/lib/src/material/app_bar.dart (hunk 1 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/app_bar.dart]: Nice.

File: packages/flutter/lib/src/material/app_bar.dart (hunk 2 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/app_bar.dart]: Also nicer.

File: packages/flutter/lib/src/material/app_bar.dart (hunk 3 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/app_bar.dart]: We could give SingleChildLayoutDelegate a const constructor...

File: packages/flutter/lib/src/rendering/sliver_persistent_header.dart [hunk 4]
```diff
@@ -265,6 +265,22 @@ abstract class RenderSliverFloatingPersistentHeader extends RenderSliverPersiste
   // direction. Negative if we're scrolled off the top.
   double _childPosition;
 
+  @protected
+  void _updateGeometry() {
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/sliver_persistent_header.dart]: I'd give this a public name since the classes are public.

**Comment by abarth** [comment]
LGTM

**Comment by Hixie** [comment]
Might make sense to add this to the demos in the contacts page on the gallery, since that demo shows the three other ways of doing app bars already.

**Comment by Hixie** [comment]
This PR decreased coverage.

**Comment by bunopus** [comment]
@HansMuller, @Hixie Is this still working now? I'm trying to achieve that (hide title on scroll and leave bottom pinned), but looks like it's doesn't work. Title is pinned as well as bottom part of SliverAppBar.
Thank you!

**Pull request merged**